### ISSUE TITLE
Unable to list contents of /var/cache/salt/minion/files/base/_usr_share_java_jdk-8u112-linux-x64.tar.gz

### ISSUE THREAD

**Issue by porunov** [saltstack-formulas/sun-java-formula#21]
Hello, 

I can't install Java 8 properly with this formula.

Here is the failed state:

```
----------
          ID: unpack-jdk-tarball
    Function: archive.extracted
        Name: /usr/share/java
      Result: False
     Comment: Unable to list contents of /var/cache/salt/minion/files/base/_usr_share_java_jdk-8u112-linux-x64.tar.gz. If this is an XZ-compressed tar archive, install XZ Utils to enable listing its contents. If it is compressed using something other than XZ, it may be necessary to specify CLI options to decompress the archive. See the documentation for details. Additional info follows:          
                                                                                                                                     
              archive location:                                                                                                      
                  /var/cache/salt/minion/files/base/_usr_share_java_jdk-8u112-linux-x64.tar.gz                                       
                                                                                                                                     
              If the source archive is a tar archive compressed using a compression type not natively supported by the tar command, then setting the 'list_options' argument may allow the contents to be listed. Otherwise, if Salt is unable to determine the files/directories in the archive, the following workaround(s) would need to be used for this state to proceed (assuming the source file is a valid tar archive):                                                                                                                    
                                                                                                                                     
              - Ownership cannot be managed without setting 'enforce_ownership_on'.                                                  
     Started: 11:20:00.855186
    Duration: 300.165 ms
     Changes:   
    Warnings: The 'tar_options' argument has been deprecated, please use
              'options' instead.
----------
          ID: create-java-home
    Function: alternatives.install
        Name: java-home
      Result: False
     Comment: One or more requisite failed: sun-java.unpack-jdk-tarball
     Changes:   

```


Here is my pillar data:

```
java_home: /usr/lib/java
java:
  source_url: http://download.oracle.com/otn-pub/java/jdk/8u112-b15/jdk-8u112-linux-x64.tar.gz
  jce_url: http://download.oracle.com/otn-pub/java/jce/8/jce_policy-8.zip
  version_name: jdk-linux-server-x64-8u112
  prefix: /usr/share/java
  dl_opts: -L

# java:version_name is the name of the top-level directory inside the tarball
# java:prefix is where the tarball is unpacked into - prefix/version_name being
#             the location of the jdk
# java:dl_opts - cli args to cURL

```

What am I doing wrong?

Sincerely,
Alexandr

**Comment by corux** [comment]
Hi Alexandr,

try to remove the `java:dl_opts` setting from your pillars.

The download of the JDK requires to accept the oracle license terms. This is already handled by the default `dl_opts` specified in the formula (see https://github.com/saltstack-formulas/sun-java-formula/blob/master/sun-java/settings.sls#L12).

Best,
Tobias

**Comment by porunov** [comment]
Hello Tobias,

I tried to remove it but it still doesn't work.. The same error happens. The thing is that tar archive seems downloaded correctly but it can not be unpacked.

Do you know what it can be else?

Sincerely,
Alexandr

**Comment by corux** [comment]
Hi Alexandr,

I just verified this formula with salt 2016.11.1 running on CentOS 7. Your pillars without the `java:dl_opts` seem to work fine.

You can try to clear the salt minion cache. Run `salt-call saltutil.clear_cache` on the minion.
The cached file `/var/cache/salt/minion/files/base/_usr_share_java_jdk-8u112-linux-x64.tar.gz` should be removed then.

Best,
Tobias

**Comment by porunov** [comment]
Thank you for help. I found out that it installs jdk in `/usr/share/java/jdk1.8.0_112/` but not in `/usr/lib/java`. Do you know why?

Sincerely,
Alexandr

**Comment by corux** [comment]
If I read the `init.sls` correctly, the path in the pillar value `java_home` (i.e. `/usr/lib/java`) will be a symlink to the extracted jdk in `java:prefix` (i.e. `/usr/share/java/jdk1.8.0_112`).

That way the `JAVA_HOME` environment variable does not need to be updated every time a new jdk gets extracted.

I'll close this issue, as your problem seems to be solved. If you have more questions, please reopen the issue.

Best,
Tobias

**Comment by porunov** [comment]
Thank you for the explanation!

Sincerely,
Alexandr

**Comment by blbradley** [comment]
I've experienced this same error on a version of the formula from mid 2016 and Salt 2016.11.4.

**Comment by blbradley** [comment]
Also on latest master and Salt 2016.11.5.

**Comment by myoung34** [comment]
@blbradley the issue behind this is that curl is getting a 404, downloading an html file, and attempting to unarchive it. Thats what started the talk on #25

**Comment by blbradley** [comment]
@myoung34 I'm at a loss for how you came to that conclusion.   🕳

**Comment by myoung34** [comment]
After reading the full error, its likely not the same. The issue i ran into was that `curl -b ...` was succeeding and archive.extracted was attempting to run on it causing a very similar error

**Comment by sroegner** [comment]
I am seeing the same issues: as if OTN changed the way their portal downloads worked all i ever get is a HTTP 404 unless I use the fully resolved redirect URL (like http://download.oracle.com/otn-pub/java/jdk/8u131-b11/d54c1d3a095b4ff2b6607d096fa80163/jdk-8u131-linux-x64.tar.gz).

**Comment by blbradley** [comment]
@sroegner What version of Salt?

**Comment by sroegner** [comment]
`salt-call 2016.11.5 (Carbon)`

**Comment by blbradley** [comment]
@sroegner I looked around and agree with your suspicion. Thanks for the report. Not sure how to resolve yet.

**Comment by blbradley** [comment]
This could be (somewhat) resolved by allowing `salt://` URLs and hosting the Java distribution on a Salt master file server.

I've done that in hadoop-formula before, but this formula currently expects the source url to be processed by curl. In hadoop-formula, I believe I removed the curl logic and gave the source url to `archive.extracted`. I could do the same here.

**Comment by sroegner** [comment]
I believe those *resolved* URLs (coded in js, so they aren't redirects) are stable and good enough to put in the source_url, I am not a big fan of prescribing the use of salt URLs (we download all of our tarballs from internal mirrors via http) to all of our (many) salt clusters which themselves are auto-provisioned

**Comment by blbradley** [comment]
@sroegner I wouldn't want a `salt://` URL as the default, but having the mechanism for it would be nice.

> good enough to put in the source_url

Good enough for the formula default? LGTM until there is a better solution.

**Comment by sroegner** [comment]
Why don't we make the salt URL a different issue and just put the new URL in place to resolve this one?

**Comment by blbradley** [comment]
Agreed!

**Comment by blbradley** [comment]
Closed in #27.
